[WIP] FEA New meta-estimator to post-tune the decision_function/predict_proba threshold for binary classifiers

glemaitre opened the pull request:
closes #8614
closes #10117
supersedes #10117 

# Description

This meta-estimator is intended to find the decision threshold which will maximize an objective metric. There are 2 use-cases:

* maximize a metric such as `balanced_accuracy_score`, `f_beta`, `f1_score`, etc. In this case, we are required to maximize the score.
* find the decision threshold for a couple of metrics for which one of the metric value will be fixed. For instance, for precision-recall, one would like the threshold maximizing the recall for a given precision or vice-versa. I think that we can support precision/recall and TPR/FPR at first.

# Additional work

* in the future, we could think of adding support for passing a cost-sensitive matrix of the size of the confusion matrix. It will allow injecting business-oriented information regarding the different classification errors done by a  learner.

glemaitre committed:
FEA add CutoffCalibration estimator

glemaitre committed:
add example

glemaitre committed:
PEP8

glemaitre committed:
add whats new entry

glemaitre committed:
mark as only working with binary data

glemaitre committed:
common tests fixes

glemaitre committed:
iter

glemaitre committed:
xxx

glemaitre committed:
iter

glemaitre committed:
pep8

ogrisel commented:
I am not sure it belongs to `sklearn.calibration`: to me the `sklearn.calibration` module is about ensuring that the model predicted value can match the empirical mean of the observed variable on a test set (on groups of samples binned by predicted values).

For instance a model that always predict 0.5 on a balanced binary classification problem is well-calibrated but not discriminative. On the contrary a model with a good ROC-AUC curve can be very discriminative (it ranks the samples in the right order most of the time) but does not necessarily predict the expected value of the mean observation correction: arbitrary strictly monotic transformation of the `predict_proba` values will preserve the curve but will break the calibration.

This definition of model calibration matches this:

https://en.wikipedia.org/wiki/Probabilistic_classification

and that:

https://blogs.sas.com/content/iml/2018/05/14/calibration-plots-in-sas.html#prettyPhoto

So to me model calibration, in the context of classification, is a property of the values returned by `predict_proba` and is therefore independent of a specific cut-off value used to make hard decisions.

If would therefore consider moving the CutoffClassifier class elsewhere, for instance in `sklearn.model_selection` as to me it has more in common with Grid-Search than probability calibration.

That being said, we could also re-scope the `sklearn.calibration` and keep Cut-off tuning there but we should be extra pedagogical about the aim of each meta-estimator to highlight they do not solve the same problem at all.

glemaitre committed:
iter

glemaitre committed:
move to model_selection

glemaitre committed:
add the missing files

glemaitre committed:
Remove current documentation

glemaitre committed:
DOC add docstring examples

glemaitre committed:
PEP8

glemaitre committed:
fix for predict proba

glemaitre committed:
add support for cost-sensitive

glemaitre committed:
revert calibration changes

glemaitre commented:
@ogrisel I think that the API starts to be stable. I added support for the cost-matrix since it is really close to the support for any metric. The cost matrix is passed using the parameter `objective_value` which, IMO, seems to be suited enough for this. I added a couple of tests to cover the general + the common tests.

I will add the documentation tomorrow by merging the previous doc and improve the blending with some examples.

At last, I think that we should rename the estimator, even more, if we support the cost-matrix. I was thinking maybe to `OptimumThresholdClassifier`.

glemaitre commented:
Even if this is still WIP (I need the documentation), @adrinjalali and @amueller might be interested to have a look at it.

glemaitre commented:
And one thing about API would be to know if we should use CV or not for selecting the threshold. If yes, should it be a single train/test split or a `cross_val_predict` (forbidding shuffle CVs), or a full `cross_validate`.

glemaitre commented:
> OptimumThresholdClassifier

Instead of `Optimum` it could be a term linked to the fine-tuning using the `objective_metric`.

ogrisel commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,339 @@
+    @staticmethod
+    def cost_sensitive_score(y_true, y_pred, cost_matrix):
+        cm = confusion_matrix(y_true, y_pred) * cost_matrix
+        return np.diag(cm) / cm.sum()
```
Why pass a full cost matrix if you then only consider the diagonal?

We could define an arbitrary business-specific utility function: such as:

- gain for true positive
- gain (negative cost) for false positives
- gain (negative cost) for false negatives
- gain for true negative (generally 0 for imbalanced classification problems).

I which case `cost_senstive_score` would be `(confusion_matrix * gain_matrix).sum() / n_samples` or alternatively `-(confusion_matrix * cost_matrix).sum() / n_samples`.

WDYT?

This work could be later extended to have parametrized cost sensitive scorer objects that could both be used for optimal cutoff selection of the final classifier of a pipeline but also used by the SearchCV object for hyper-parameter tuning of the full pipeline in general.

glemaitre replied:
> Why pass a full cost matrix if you then only consider the diagonal?

`cm.sum()` will consider off-diagonal. If you penalize differently false-positive and false-negative, the cost will change in this case, isn't it?

ogrisel replied:
Indeed. This is a really hard to interpret how gains and costs are counteracts one another. I prefer my simple purely additive parametrization unless you can show me that yours make more sense.

glemaitre replied:
I went a bit through the paper of Elkan: http://web.cs.iastate.edu/~honavar/elkan.pdf
In section 2, there is an analytic formulation of the optimal threshold. However, it requires to use `predict_proba` (which could be fine since `method="auto"` by default?).
In this case (and I understand correctly), it means that you can compute the threshold just by looking at the cost matrix.

glemaitre replied:
Otherwise, your additive parametrization if fine with me, it makes sense.

ogrisel replied:
The Elkan paper also mentions the case of non-constant costs, that is costs that depends on the value of an input feature. This is practically very useful, probably much more useful than constant costs. Maybe we should re-scope this whole cost matrix thingy for another PR though. It will be much easier to implement once we have a clean API to deal with feature names.

ogrisel reviewed (commented):
w.r.t. CV vs single split I think the main point from @marctorrellas in https://github.com/scikit-learn/scikit-learn/pull/10117#issuecomment-590781742 is "refit" vs "no refit". When you CV you implicitly refit on the full training model. Or alternatively you would have to keep the k fitted models on each k-splits and take majority vote afterwards but that feels weird.

If you do a single split, then you have the opportunity to keep the estimator fitted on the train side of the split and then the selected threshold matches this specific fitted model. If you refit on the full training set, then the selected threshold is no longer specifically tuned for this fitted model. I think we should do some experiments to drive this discussion further: choose a dataset, materialize several train / validation splits and for each of them draw the ROC curves and highlight the thresholds with optimal f1score or balanced accuracy. 

w.r.t. renaming I don't really like optimum `OptimumThresholdClassifier`. Maybe `TunedCutoffClassifier`? But I think `CutoffClassifier` was a fine name.

ogrisel commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,339 @@
+        The classifier, fitted or not fitted, from which we want to optimize
+        the decision threshold used during `predict`.
+
+    objective_metric : {"tpr", "tnr"}, ndarray of shape (2, 2) or callable, \
```
If should also include `"f1"` and `"balanced_accuracy"` and maybe others such as `"fbeta"` if we also include a new dedicted "beta" constructor param.

glemaitre replied:
I was thinking to let the user pass a skelarn.metric callable. It makes it easy because you will never need a `objective_value` for this case and only need an `objective_value` when the a `str` is provided because we will work with predefined search.

ogrisel replied:
I would still make it possible to pass strings for metrics that can be efficiently computed from the output of _binary_clf_curve. This private tool could actually be extended to `_binary_clf_confusion_matrix_curve` that would compute the CM for each threshold.

glemaitre replied:
> "fbeta" if we also include a new dedicated "beta" constructor param

This could be pass to `objective_metric_param` indeed.

glemaitre replied:
> I would still make it possible to pass strings for metrics that can be efficiently computed from the output of _binary_clf_curve. This private tool could actually be extended to _binary_clf_confusion_matrix_curve that would compute the CM for each threshold.

OK. I will first make the documentation with the naive approach and then, I will improve on the efficiency.

ogrisel reviewed (commented):
We also need an example, e.g. on a artificially imbalanced version of Adult Census with one hot encoded categorical features, scaled numerical features and Logistic Regression possibly with `class_weight="balanced"`.

Then plot the ROC and PR curves and add the points for the model with the default cutoff (0.5 predict_proba) and the model with the best balanced accuracy and the model with the best f1_score.

ogrisel commented:
Also it would be great to find the model with the best recall for a given precision constraint (e.g. precision >= 0.3) but not all precision levels are feasible. So this would make lead to models that could raise a `ValueError` or `FeasibilityError` at fit time... I am not sure we want this. as it the case for recall aka TPR.

On the other hand, all recall / TPR values are feasible, so we might want to have is given a recall / TPR constraint, find the cutoff that leads to the best precision, but this is not easy to specify with the current set constructor parameters.

Finally, if we treated the default cutoff as a regular parameter for any classifier one could also use SearchCV objects to tune it but that would be a very inefficient thing to do (retraining models for nothing). Maybe it's worth highlighting this aspect in the documentation.

glemaitre committed:
start doc

ogrisel commented:
I had not read the discussion that already happened in the collapsed discussion of the previous PR two years ago:

https://github.com/scikit-learn/scikit-learn/pull/10117#pullrequestreview-121628080

Indeed unless we do not have a clear academic reference to guide us w.r.t. discussion I would rather spend time on a notebook conducting experiments with curves and scores on unbalanced classification problems with error bounds to highlight what we should do.

I would therefore put the developments of this PR on hold until we have a good understanding on the CV + refit vs single split without refit debate.

marctorrellas commented:
Thanks for the PR @glemaitre . My proposal for the name is `TunedThresholdClassifier`. I don't like Optimal, and I think Threshold is easier to understand than Cutoff, but that's really personal preference :)

Let me explain my understanding about cv, refit, etc more precisely

* If you want to ESTIMATE PERFORMANCE of a classifier would be when tuning the threshold, what you would do is a nested CV, where
       * In the simplest case you would just have train, dev, and test. So fit parameters on train, fit the threshold (aka hyperparam) on dev, evaluate on test
       * In a more complex setting, you would do a Kfold nested with another Kfold. With a similar idea

* If you want to PRODUCE A MODEL, you split in train and dev only, fit params on train, fit hparams (threshold) on dev. Then you quote the performance as the estimate in the section above

* Common misconceptions (according to my understanding)
     * Do the first task with non-nested Kfold, and then decide that the best threshold is the average of thresholds. I guess you could justify this as considering the thresholds you have as observations of a random variable, and assuming that the average gives you the mean, which should be decent for any classifier. In some cases this might give decent results, especially if thresholds tend to be similar, but for sure it's not the optimal threshold (at least not for a specific training dataset, maybe for an average random sample of a training dataset...)
     * Quote performance on a dev set as the performance of the classifier. This is wrong as you select the threshold to max performance on it, so you should evaluate performance in a separate unseen dataset

amueller commented:
I would rather put it into ``compose`` than have a new module tbh.

glemaitre commented:
Do you the _prediction.py module?

 

                                                                                                                                      
 

 
     

Sent from my phone - sorry to be brief and potential misspell.

glemaitre committed:
iter

glemaitre commented:
So I run experiments there regarding the cross-validation: https://github.com/glemaitre/cutoff_experiments/blob/master/cutoff_experiments.ipynb

It is a bit counter-intuitive (at least for me) but a single split without a refit is actually more stable than picking the threshold corresponding the best mean average score even on small dataset size.

ping @ogrisel any idea what to look at to go further.

glemaitre commented:
Actually my bad, I did not see that my box plot excluded an outlier. So actually the cross-validation is more stable with small sample size and really similar with large sample size

glemaitre commented:
(This section is in progress)

## Summary

So here is a summary of the different experiments that I carry out. We wanted
to compare two approaches to find the cutoff threshold:

1. Make a cross-validation and get the score for each fold. Find the cutoff
   threshold maximizing the mean score across fold. Refit a model on the full
   training set. Evaluate the model using the cutoff threshold on the left-out
   testing set.
2. Split the training set into train/validation split. Fit the model on the
   training data and find the cutoff threshold maximizing the score on the
   validation set. Do not refit the model. Evaluate the model using the cutoff
   threshold on the left-out testing set.

## Experiments

We can consider to the following parameters to have an impact on the two
methods presented above:

* the size of the dataset;
* the frequency ratio of the different classes;
* the metric/score to be optimized.

To make the evaluation, we are going to use:

* a plot of the metric/score (on each fold when applicable) for different
  cutoff threshold;
* ROC and PR curves of the model and the threshold found;

### Effect of the sample size

#### 500 samples

![cv_fold_500](https://user-images.githubusercontent.com/7454015/76968721-a4f02a80-6929-11ea-9f0d-965803a0d8b5.png)
![cv_roc_pr_500](https://user-images.githubusercontent.com/7454015/76968719-a4579400-6929-11ea-916d-4e90b52e4336.png)

![single_roc_pr_500](https://user-images.githubusercontent.com/7454015/76968716-a3befd80-6929-11ea-93c5-ce26d9184d3f.png)

![comparison_500](https://user-images.githubusercontent.com/7454015/76968713-a3266700-6929-11ea-9d59-8513a2af8af1.png)

Small sample size induces a large variation for each fold of the CV. Taking the mean best score is slightly attenuating this issue. We can see in the last comparison plot that the variance across several try is reducing slightly.

#### 5000 samples

![cv_fold_5000](https://user-images.githubusercontent.com/7454015/76968711-a3266700-6929-11ea-99cb-46270a0c3288.png)
![cv_roc_pr_5000](https://user-images.githubusercontent.com/7454015/76968705-a28dd080-6929-11ea-9d74-984fa0b2dbef.png)

![single_roc_pr_5000](https://user-images.githubusercontent.com/7454015/76968700-a15ca380-6929-11ea-990a-d75daf56d9d5.png)

![comparison_5000](https://user-images.githubusercontent.com/7454015/76968697-a0c40d00-6929-11ea-8750-17cb5ccaa3a5.png)

#### Entire dataset

### Effect of class imbalance

### Effect of the metric

PGryllos commented:
@glemaitre thx for summarising your experiments here! 

> In short, small sample size induces a larger variation making the cv procedure slightly more robust.
do you mean slightly less robust? 

on the cv approach, do you select the best threshold across folds or take the mean best threshold?

glemaitre commented:
> on the cv approach, do you select the best threshold across folds or take the mean best threshold?

I take the threshold corresponding the mean best score computed across folds as proposed by @amueller

PGryllos commented:
Makes sense, I guess the idea of averaging the thresholds got dropped due to lack of literature suggesting it is meaningful?

It seems that the two approaches are similar for balanced datasets whereas the cv approach allows for better thresholds when the dataset is very imbalanced if I am reading this right?

Do you think it makes sense to see if the results are stable across different types of classifiers (e.g. CART models)?

glemaitre commented:
> Makes sense, I guess the idea of averaging the thresholds got dropped due to lack of literature suggesting it is meaningful?

yes and it seems kinda of wrong. I think that it on small sample size we might see that it should get worth.

> Do you think it makes sense to see if the results are stable across different types of classifiers (e.g. CART models)?

The plan would be to run the experiments on another dataset with an HistGradientBoosting. But at least the first batch of experiments can already give some insights.

ogrisel commented:
I agree that on average the mean test score of the "best cutoff based on mean cv score + refit" strategy is not significantly better than the "single split / no refit" strategy but the former seems to have less variance.

I think it makes sense to implement both strategies. I am not sure which should be the default. Maybe the CV-based strategy is safest.

For the single split strategy, it would be nice to also make it possible for the user to pass a predefined validation set. Although I am not sure if it should be passed as constructor param or fit_param. We have no standard way to do this in scikit-learn as far as I know and this sounds like a complex API design issue so let's postpone pre-defined validation sets for later.

glemaitre commented:
Here come the new experiment on Ames dataset (which binarized to transform the regression problem into a classification problem). I binarize by taking the 90th percentile as a  threshold.


![cv_fold_ames](https://user-images.githubusercontent.com/7454015/77078188-89068a80-69f6-11ea-8bfa-4e7b41d9aec7.png)
![cv_roc_pr_ames](https://user-images.githubusercontent.com/7454015/77078183-86a43080-69f6-11ea-9d81-8c00759ba04d.png)

![single_roc_pr_ames](https://user-images.githubusercontent.com/7454015/77078180-860b9a00-69f6-11ea-83b2-7cea36d2f3f1.png)

![comparison_ames](https://user-images.githubusercontent.com/7454015/77078178-84da6d00-69f6-11ea-8710-c8309444f7db.png)

glemaitre commented:
So we get something similar than in the previous experiment with a different classifier and dataset (more variance for the single split).

ogrisel commented:
Thanks very much @glemaitre for running those experiments. So indeed I think I would be in favor of the "best cutoff on mean CV score + refit" as the default strategy (with cv=5 as a default value for consistency with the rest of scikit-learn) and allow for the "single split no refit" strategy for users that are limited by computational constraints on datasets with a large enough number of samples.

glemaitre committed:
Merge remote-tracking branch 'origin/master' into is/10117

glemaitre committed:
skip test

glemaitre committed:
iter

glemaitre committed:
remove unsued code

glemaitre committed:
docstring

glemaitre committed:
pep9

ntlghb commented:
Thank you @glemaitre for working on this and for considering to include the cost matrix!

In my experience, the method you use can impact performance significantly, with certain classifiers producing a much wider range of unique thresholds than others. Pre-training methods and Ensembles that incorporate those produce much wider range of unique thresholds, in my experience.
Below are some illustrations from experients I did for my paper. These are raw (uncalibrated) thresholds and associated balanced accuracy scores from 20 iterations. Each iteration selects a threshold on the validation set which is the applied to the disjoint test set to look up the balanced accuracy score.

**CART**
![all_thresh_dt_none_20](https://user-images.githubusercontent.com/42805270/78108087-f94ddc80-73f6-11ea-8d24-88e5a29aee18.png)

**Adaboost (base estimator = CART, max_depth=1)**
![all_thresh_adab1_none_20](https://user-images.githubusercontent.com/42805270/78108281-5cd80a00-73f7-11ea-99a3-d6c3dcaed697.png)

**Balanced random forest**
![all_thresh_bal_rf_none_20](https://user-images.githubusercontent.com/42805270/78108548-e8ea3180-73f7-11ea-9912-602df4052865.png)

**RUSboost**
![all_thresh_rusboost_none_20](https://user-images.githubusercontent.com/42805270/78109207-1aafc800-73f9-11ea-91f4-1fd1bca6d4a7.png)

Regarding the costmatrix scorer, while any matrix can be transformed to one that has zeros off the diagonal (like Elkan suggests), not all applications are interested in assigning zero cost to correct classifications. Typical domain is fraud detection, for example see cost matrix definition in
 https://www.semanticscholar.org/paper/ROC-graphs-with-instance-varying-costs-Fawcett/b3d913726aa8b139076fd406b1585250ea6910ac

May I suggest to allow the user to pass a  n_samples x 4  cost matrix of record-dependent costs instead of defaulting to constant misclassification costs and only summing the error costs?

NicolasHug commented on doc/whats_new/v0.23.rst:
```diff
@@ -328,6 +328,12 @@ Changelog
   :user:`Arie Pratama Sutiono <ariepratama>`
 
+- |MajorFeature| :class:`model_selection.CutoffClassifier` calibrates the
+  decision threshold function of a classifier by maximizing a binary
```
```suggestion
  decision threshold function of a binary classifier by maximizing a
```

NicolasHug commented on doc/whats_new/v0.23.rst:
```diff
@@ -328,6 +328,12 @@ Changelog
 
+- |MajorFeature| :class:`model_selection.CutoffClassifier` calibrates the
+  decision threshold function of a classifier by maximizing a binary
+  classification metric through cross-validation.
```
```suggestion
  classification metric through cross-validation.
  That decision threshold is then used to convert a probability or the output of the decision function into a predicted class.
```

NicolasHug commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,462 @@
+        The classifier, fitted or not fitted, from which we want to optimize
+        the decision threshold used during `predict`.
+
+    objective_metric : callable or {"tpr", "tnr"} \
+            default=balanced_accuracy_score
```
Should this also accept strings? e.g. for the mapping in https://scikit-learn.org/stable/modules/model_evaluation.html#balanced-accuracy-score

glemaitre replied:
True, I don't see yet why we could not use the internal scorer.

NicolasHug commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,462 @@
+    """Decision threshold calibration for binary classification.
+
+    Estimator that calibrates the decision threshold (cutoff point) that is
+    used for prediction. The methods for picking cutoff points make use of
```
```suggestion
    used for converting a probability or a decision function into a predicted class. The methods for picking cutoff points make use of
```

NicolasHug replied:
By "The methods" you mean the CV procedure? It slightly ambiguous with the `method` parameter.

NicolasHug commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,462 @@
+        want to find the decision threshold when `objective_metric` is equal to
+        `"tpr"` or `"tnr"`.
+
+    method : {"auto", "decision_function", "predict_proba"}, default="auto"
```
This is typically called `response_method` in the PDP, precision_recall and ROC curves

NicolasHug commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,462 @@
+        decision function for which we want to find a threshold. It can be:
+
+        * if `"auto"`, it will try to invoke, for each classifier,
+          `"decision_function` or `"predict_proba"` in that order.
```
Any reason for this order? Intuitively, it would make more sense to me to tune the threshold on `predict_proba` first.

glemaitre replied:
I think that it was proposed in this order. Having recently worked on the scorer, they actually try `decision_function` first.
I don't know if most of the time when `decision_function` is available, it provides a native function from the algorithm and the probability estimate is just an adaptation of the decision function?

NicolasHug commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,462 @@
+          If the method is not implemented by the classifier, it will raise an
+          error.
+
+    n_threshold : int, default=1000
```
```suggestion
    n_thresholds : int, default=1000
```

NicolasHug commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,462 @@
+    cv : int, float, cross-validation generator, iterable or "prefit", \
+            default=None
+        Determines the cross-validation splitting strategy used in
+        `cross_val_predict` to train classifier. Possible inputs for cv are:
```
There's no use of `cross_val_predict` ?

NicolasHug commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,462 @@
+    Parameters
+    ----------
+    base_estimator : estimator object
+        The classifier, fitted or not fitted, from which we want to optimize
```
```suggestion
        The classifier, fitted or not fitted, for which we want to optimize
```

NicolasHug commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,462 @@
+    objective_metric_params : dict, default=None
+        Some extra parameters to pass to `objective_metric`.
+
+    objective_value : float, default=None
+        The value associated with the `objective_metric` metric for which we
+        want to find the decision threshold when `objective_metric` is equal to
+        `"tpr"` or `"tnr"`.
```
What does None do if tpr or tnr is passed? Looking briefly I'm not sure it raises an error?

NicolasHug commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,462 @@
+        cross-validation strategies that can be used here.
+
+    refit : "auto" or bool, default="auto"
+        Whether or not to refit the classifier on the entire training set once
+        the decision threshold has been found. By default, `refit="auto"` is
+        equivalent to `refit=False` when `cv` is a float number using a single
+        shuffle split or `cv="prefit"` otherwise `refit=True` in all other
+        cases. Note that forcing `refit=False` on cross-validation having more
+        than a single split will raise an error. Similarly, `refit=True` in
+        conjunction with `cv="prefit"` will raise an error.
```
I understand the logic, but is it overkill? I feel like this should follow the same simple logic as the Grid Search estimators

glemaitre replied:
If we choose the way of the GridSearchCV, `refit=False` will raise most of the time an error.
If we choose `refit=True`, it will refit when a user ask for a single split while it would be the case where you want to be efficient most probably (because you have a large dataset).
So the "auto" switch from one to another and seem a good default to me. If I recall, we made this choice with @ogrisel

NicolasHug commented on sklearn/model_selection/_prediction.py:
```diff
@@ -0,0 +1,462 @@
+    n_jobs : int, default=None
+        The number of jobs to run in parallel all `estimators` `fit`.
+        `None` means 1 unless in a `joblib.parallel_backend` context. -1 means
+        using all processors. See Glossary for more details.
```
Needs a link ;)

NicolasHug reviewed (commented):
Cool stuff!

made a first quick pass on the docs + some questions

marctorsoc commented:
what's the state of this?

glemaitre commented:
We are having a sprint currently so it is on hold for this week but I am planning to work on it from next week.

glemaitre committed:
TST wip

glemaitre committed:
TST wip

glemaitre committed:
TST wip

glemaitre committed:
TST PEP8 + comments

glemaitre committed:
TST force to use predict_proba as well

glemaitre committed:
DOC add whats + PEP8

glemaitre committed:
TST add some tolerance since the average of squared in diff ordered

lorentzenchr commented:
Is it also possible to set the threshold for `predict_proba` manually?
Use Case: The probability predictions of a well calibrated `LogisticRegression` already represent a measure of (un)certainty of the predictions, i.e. the accuracy.

As to the name, like @marctorsoc, I'd prefer the term *threshold* over *cutoff* in this context.

marctorsoc commented:
I think it'd be great to have a hyperparam decision threshold so that one can initialise the estimator with it, eg. after hyperopt

glemaitre committed:
STY add better error message and refactor code

glemaitre committed:
fix

glemaitre committed:
fix

glemaitre committed:
Merge remote-tracking branch 'origin/master' into is/10117

glemaitre committed:
Update sklearn/metrics/tests/test_score_objects.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

glemaitre committed:
add test for PredictScorer

glemaitre committed:
apply olivier suggestions

glemaitre committed:
use list

glemaitre committed:
Merge remote-tracking branch 'origin/master' into is/10117

glemaitre committed:
fix

glemaitre committed:
Merge remote-tracking branch 'glemaitre/is/scorer_pos_label' into is/10117

glemaitre committed:
fix

glemaitre committed:
PEP8

glemaitre committed:
Merge branch 'is/scorer_pos_label' into is/10117

glemaitre committed:
finally passing?

rth commented:
Looking forward to this! 

Is there any possibilities to generalize this to multi-class cases? I haven't really studied the question, but for instance for models that have a `predict_proba` method, instead of defining 1 threshold, I imagine it would be somewhat equivalent to define the prediction as `np.argmax(w * proba, axis=1)`, where w is a vector of shape (n_classes) to be optimized with some additional constraint, say sum(w) = 1, then use the multi-class averaged f1 etc scores?

marctorsoc commented:
> Looking forward to this!
> 
> Is there any possibilities to generalize this to multi-class cases? I haven't really studied the question, but for instance for models that have a `predict_proba` method, instead of defining 1 threshold, I imagine it would be somewhat equivalent to define the prediction as `np.argmax(w * proba, axis=1)`, where w is a vector of shape (n_classes) to be optimized with some additional constraint, say sum(w) = 1, then use the multi-class averaged f1 etc scores?

IMHO, let's have it for binary first, and then we can move from that. This PR has been here for a long time with a lot of discussion, so better not to increase the scope of the task

glemaitre commented:
> what's the state of this @glemaitre ? do you need help with it?

I am fixing something related to the scorer in #18114 
I want to have this first to be sure that we have a clean API regarding the scoring passed.

lorentzenchr commented:
Just to set the reference, it would be nice to be able to also close #4813 with this meta-estimator.

adrinjalali commented:
Not sure if there's time to get this in, since it also seem to depend on #18589. I'll leave it here, but isn't a blocker for the release.

UnixJunkie commented:
Supporting MCC as one of the metrics would be nice.

lorentzenchr commented:
superseded by https://github.com/scikit-learn/scikit-learn/pull/26120

skanskan commented:
Can this be used at the same time than GridSearchCV?
Is it the same than using TunedThresholdClassifierCV with refit=True?
